### ISSUE TITLE
Reinstate simd_cast but implement using rebind

### ISSUE THREAD

**Issue by danieltowner** [mattkretz/std-simd-feedback#34]
Real-world use of rebind_simd_t suggests that simd_cast is a concise and readable way of indiciating what you mean, while rebind_simd_t is a bit verbose.

**Comment by mattkretz** [comment]
I'd like to see a paper to SG6 proposing a general `value_type` / representation type cast.
* `std::value_cast<double>(1)`: `int` to `double`
* `std::value_cast<int>(simd)`: `simd<T>` to `rebind_simd_t<int, simd<T>>`
* maybe `std::value_cast<float>(5min)`: `duration<int, std::ratio<60>>` to `duration<float, std::ratio<60>>`
* `std::value_cast<double>(1.f + 1.if)`: `complex<float> to complex<double>`
* ... esp. calling out to papers on numeric library types

**Comment by mattkretz** [comment]
@mpusz FYI. I guess such a cast would be interesting for mp-units as well. It would be equivalent to your `quantity_cast` overload for casting the representation type.

**Comment by mpusz** [comment]
Yes, we call it a `representation_cast` in mp-units V2 framework for now but `value_cast` is also a good name. However, in our case it will cast not only an underlying representation type but also will be used to change the unit as such conversion may also change/truncate the value. To make it even more entertaining, Unit is provided by value (NTTP) and representation type as a type parameter kind.

For example:

```cpp
auto s1 = representation_cast<int>(speed);
auto s2 = representation_cast<km / h>(speed);
```

**Comment by mattkretz** [comment]
Assuming WG21 adds a `std::value_cast` with a specific goal to allow UDTs to reuse it. Then there would be some mechanism to call a function defined by mp-units that implements the cast.

I wrote a quick prototype: https://godbolt.org/z/M6e5q9ao7.

Who wants to run with it and write a paper? :wink:

**Comment by danieltowner** [comment]
> Who wants to run with it and write a paper? 😉

This idea came up again this week at (https://wiki.edg.com/bin/view/Wg21sofia2025/P3445), so while it was fresh in my mind I had a go: https://isocpp.org/papers/form/14571

**Comment by mattkretz** [comment]
👍 Your second link isn't the link to the paper. I'd like to read it :-)

**Comment by danieltowner** [comment]
Doh. https://isocpp.org/files/papers/D3750R0.html. Also available in our git collaboration repo.

My original desire for the feature in simd was to change the `value_type` of the simd container to something else, and having that function be generalised to handling a change in any underlying type which had a `value_type` (array, vector, etc.) is my main motivation. To enable simd-generic programming the `value_cast` also needs to be able to deal with scalar types, but in the conversation above you want `value_cast` to be able to convert UDTs, unit-based values, and so on as well, and I don't think I have done that use enough justice in the first draft of the paper.

I added @mpusz and @mattkretz as co-authors because you both contributed to the discussion above that resulted in the paper.

**Comment by mpusz** [comment]
Hi @danieltowner 👋 

Did you mean 'lengthAsInt'?

Regarding the implementation experience: https://github.com/mpusz/mp-units/blob/master/src%2Fcore%2Finclude%2Fmp-units%2Fframework%2Fvalue_cast.h

Please note that we have multiple versions of `value_cast()`. Some of them even take two template arguments (unit and a representation type):

```cpp
using my_length = quantity<si::metre, int>;

auto q1 = value_cast<km>(42 * m);
auto q2 = value_cast<int>(3.14 * m);
auto q3 = value_cast<int, km>(3.14 * m);
auto q4 = value_cast<km, int>(3.14 * m);
auto q5 = value_cast<my_length>(3.14 * m);
```
